### PR TITLE
Better handling of filters when checking resources in local collections

### LINKED ISSUE #4439: Interaction agents causing issues saving
**Describe the bug**
Not entirely sure what the issue is but when adding an agent to an interaction it lets you save but after it saves the buttons then turn red and it won't let you do anything else. 

**To Reproduce**
1. Create a new disposal 
2. Add disposal agent 
3. Click save

**Expected behavior**
It should save without an issue 

**Screenshots**
Screenshot 2024-01-19 142354

**Desktop:**
 - OS: Windows 11
 - Browser: Chrome
 - Specify 7 Version: edge

**Database Name:** lsumz mammals

### PR DESCRIPTION
**Pull request by melton-jason**
Fixes #4439 

### Checklist

- [x] Self-review the PR after opening it to make sure the changes look good
      and self-explanatory (or properly documented)
- [x] Add automated tests
- [x] Add relevant issue to release milestone

### Testing instructions
Ensure uniqueness rules which exist as a `to-many` relationship on another form (InteractionAgents on the Interaction form, Determination/Preparation on CollectionObject, etc.) are properly being checked and behave correctly.

### CODE REVIEW

**Review by CarolineDenis — APPROVED** [review]

**Review by bronwyncombs — APPROVED** [review]
Saving as expected! Uniqueness rules for to-many relationships are working as expected 
Screenshot 2024-01-22 at 9 57 08 AM
Screenshot 2024-01-22 at 9 58 55 AM
Screenshot 2024-01-22 at 9 57 32 AM

**Review by pashiav — APPROVED** [review]
Looks good! Buttons are functional after save.
Screen Shot 2024-01-22 at 3 19 32 PM

File: specifyweb/frontend/js_src/lib/components/DataModel/businessRules.ts [hunk 1]
```diff
@@ -234,65 +233,158 @@ export class BusinessRuleManager<SCHEMA extends AnySchema> {
...
+
+    const generateFilters = async (
+      resource: SpecifyResource<AnySchema>,
+      fieldNames: RA<string>
+    ): Promise<
+      IR<{
+        readonly id: number | undefined;
+        readonly cid: string | undefined;
+        readonly value: number | string | null | undefined;
+      }>
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/lib/components/DataModel/businessRules.ts:290]: this type seems to be repeated several times in this file - want to create a `type` for it?

File: specifyweb/frontend/js_src/lib/components/DataModel/businessRules.ts [hunk 2]
```diff
@@ -234,65 +233,158 @@ export class BusinessRuleManager<SCHEMA extends AnySchema> {
       ),
     };
 
-    const hasSameValues = (
+    const getFieldValue = async (
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/lib/components/DataModel/businessRules.ts:236]: are these functions using any variables from the scope in which they are defined (the `uniqueIn` method)?
if not, perhaps you can turn them into functions outside the class, or their own methods?
`uniqueIn` is kind of a hot-path function, so re-creating large arrow functions in here could hurt the performance (makes the job of the optimizer harder)

**Review by maxpatiiuk — COMMENTED** [review]
yay, more tests ❤️

**Pull request merged**